File: packages/kits/NeuronKit/Tests/NeuronKitTests/EstateManifestPolicyStoreTests.swift
// EstateManifestPolicyStoreTests.swift
//
// F6 / ADR-020: the manifest-backed policy stores persist dreaming/maintenance
// policy, bandit, and daemon cycle state THROUGH the public substrate interface
// (GeniusLocusKit.estate(for:) → Estate.meta/setMeta). These tests prove the
// round-trip through a live estate. Cross-restart durability of the underlying
// manifest table is proven separately by LocusKit's EstateTests reopen test.

import Foundation
import Testing
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import SubstrateTypes
@testable import NeuronKit

@Suite("EstateManifestPolicyStore — manifest-backed daemon persistence")
struct EstateManifestPolicyStoreTests {

    private func makeKit() async throws -> (GeniusLocusKit, EstateHandle) {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "manifest-store-test")
        let config = EstateConfiguration(estateID: UUID(), backend: .inMemory)
        let storage = InMemoryStorage(configuration: config)
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner)
        return (kit, handle)
    }

    // MARK: - Dreaming

    @Test("Dreaming policy round-trips through the estate manifest")
    func dreamingPolicyRoundTrip() async throws {
        let (kit, handle) = try await makeKit()
        let store = EstateManifestDreamingPolicyStore(handle: handle, kit: kit)

        // Absent before first save.
        #expect(try await store.loadPolicy() == nil)

        let policy = DreamingPolicy(
            minSuccessRate: 0.55, minConfidence: 0.8, minAttempts: 5,
            tickIntervalMs: 45_000, eventObservationThreshold: 2)
        try await store.savePolicy(policy)
        #expect(try await store.loadPolicy() == policy)
    }

    @Test("Dreaming bandit round-trips through the estate manifest")
    func dreamingBanditRoundTrip() async throws {
        let (kit, handle) = try await makeKit()
        let store = EstateManifestDreamingPolicyStore(handle: handle, kit: kit)

        #expect(try await store.loadBandit() == nil)

        var bandit = SolverBandit()
        bandit.observe(arm: .timer, reward: 1.0)
        bandit.observe(arm: .event, reward: 0.0)
        try await store.saveBandit(bandit)
        #expect(try await store.loadBandit() == bandit)
    }

    @Test("Dreaming daemon state round-trips through the estate manifest")
    func dreamingDaemonStateRoundTrip() async throws {
        let (kit, handle) = try await makeKit()
        let store = EstateManifestDreamingPolicyStore(handle: handle, kit: kit)

        #expect(try await store.loadDaemonState() == nil)

        let state = DreamingDaemonState(
            lastTickAt: Date(timeIntervalSince1970: 1_700_000_000),
            proposedKeys: ["a|b", "c|d"],
            lastReindexVocab: 1_234,
            consolidated: ["a|b": 0.88, "c|d": 0.42],
            cycleCount: 7)
        try await store.saveDaemonState(state)
        #expect(try await store.loadDaemonState() == state)
    }

    // MARK: - Maintenance

    @Test("Maintenance policy round-trips through the estate manifest")
    func maintenancePolicyRoundTrip() async throws {
        let (kit, handle) = try await makeKit()
        let store = EstateManifestMaintenancePolicyStore(handle: handle, kit: kit)

        #expect(try await store.loadPolicy() == nil)

        let policy = MaintenancePolicy()
        try await store.savePolicy(policy)
        #expect(try await store.loadPolicy() == policy)
    }

    @Test("Maintenance daemon state round-trips through the estate manifest")
    func maintenanceDaemonStateRoundTrip() async throws {
        let (kit, handle) = try await makeKit()
        let store = EstateManifestMaintenancePolicyStore(handle: handle, kit: kit)

        #expect(try await store.loadDaemonState() == nil)

        let state = MaintenanceDaemonState(
            lastTickAt: Date(timeIntervalSince1970: 1_700_000_500),
            lastAuditCheckAt: Date(timeIntervalSince1970: 1_700_000_400),
            proposedKeys: ["decay:room-1", "tombstone:row-9"],
            cycleCount: 3)
        try await store.saveDaemonState(state)
        #expect(try await store.loadDaemonState() == state)
    }
}

File: packages/kits/NeuronKit/rust/tests/estate_manifest_policy_store_tests.rs
//! F6 / ADR-020: the manifest-backed policy stores persist dreaming/maintenance
//! policy and daemon cycle state THROUGH the substrate's public KV surface
//! (DrawerStore::get_meta/set_meta). These tests prove the round-trip through a
//! live in-memory estate store. Mirrors Swift `EstateManifestPolicyStoreTests`.

use std::collections::BTreeMap;
use std::sync::Arc;

use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;

use neuron_kit::dreaming_cycle::{DreamingDaemonState, DreamingPolicy, DreamingPolicyStore};
use neuron_kit::estate_manifest_policy_store::{
    EstateManifestDreamingPolicyStore, EstateManifestMaintenancePolicyStore,
};
use neuron_kit::maintenance_cycle::{
    MaintenanceDaemonState, MaintenancePolicy, MaintenancePolicyStore,
};
use neuron_kit::solver_bandit::{DreamingTriggerMode, SolverBandit};

const NOW: i64 = 1_700_000_000;

fn store() -> Arc<dyn DrawerStore> {
    Arc::new(InMemoryDrawerStore::new(NOW, None).unwrap())
}

#[test]
fn dreaming_policy_round_trips_through_manifest() {
    let s = store();
    let mut dstore = EstateManifestDreamingPolicyStore::new(Arc::clone(&s));
    assert_eq!(dstore.load_policy(), None, "absent before first save");

    let mut policy = DreamingPolicy::default();
    policy.min_confidence = 0.8;
    policy.tick_interval_ms = 45_000;
    dstore.save_policy(policy);
    assert_eq!(dstore.load_policy(), Some(policy));
}

#[test]
fn dreaming_bandit_round_trips_through_manifest() {
    let s = store();
    let mut dstore = EstateManifestDreamingPolicyStore::new(Arc::clone(&s));
    assert_eq!(dstore.load_bandit(), None, "absent before first save");

    let mut bandit = SolverBandit::default();
    bandit.observe(DreamingTriggerMode::Timer, 1.0);
    bandit.observe(DreamingTriggerMode::Event, 0.0);
    dstore.save_bandit(bandit.clone());
    assert_eq!(dstore.load_bandit(), Some(bandit));
}

#[test]
fn dreaming_daemon_state_round_trips_through_manifest() {
    let s = store();
    let mut dstore = EstateManifestDreamingPolicyStore::new(Arc::clone(&s));
    assert_eq!(dstore.load_daemon_state(), None);

    let mut consolidated = BTreeMap::new();
    consolidated.insert("a|b".to_string(), 0.88_f32);
    consolidated.insert("c|d".to_string(), 0.42_f32);
    let state = DreamingDaemonState {
        last_timer_fire_epoch_secs: Some(1_700_000_000.0),
        proposed_keys: vec!["a|b".to_string(), "c|d".to_string()],
        last_reindex_vocab: 1_234,
        consolidated,
        cycle_count: 7,
        co_recall_counts: BTreeMap::new(),
        last_theta_run_epoch_secs: None,
        last_beta_run_epoch_secs: None,
        last_omega_run_epoch_secs: None,
    };
    dstore.save_daemon_state(state.clone());
    assert_eq!(dstore.load_daemon_state(), Some(state));
}

#[test]
fn maintenance_policy_round_trips_through_manifest() {
    let s = store();
    let mut mstore = EstateManifestMaintenancePolicyStore::new(Arc::clone(&s));
    assert_eq!(mstore.load_policy(), None);

    let policy = MaintenancePolicy::default();
    mstore.save_policy(policy);
    assert_eq!(mstore.load_policy(), Some(policy));
}

#[test]
fn maintenance_daemon_state_round_trips_through_manifest() {
    let s = store();
    let mut mstore = EstateManifestMaintenancePolicyStore::new(Arc::clone(&s));
    assert_eq!(mstore.load_daemon_state(), None);

    let state = MaintenanceDaemonState {
        last_fire_epoch_secs: Some(1_700_000_500.0),
        last_audit_check_epoch_secs: Some(1_700_000_400.0),
        proposed_keys: vec!["decay:room-1".to_string(), "tombstone:row-9".to_string()],
        cycle_count: 3,
    };
    mstore.save_daemon_state(state.clone());
    assert_eq!(mstore.load_daemon_state(), Some(state));
}

/// Two stores over the SAME backing DrawerStore observe each other's writes —
/// the persistence is in the shared manifest table, not store-local memory
/// (the cross-restart property, simulated by a second store instance).
#[test]
fn daemon_state_persists_across_store_instances() {
    let s = store();
    let mut first = EstateManifestDreamingPolicyStore::new(Arc::clone(&s));
    let state = DreamingDaemonState {
        last_timer_fire_epoch_secs: Some(42.0),
        proposed_keys: vec!["x|y".to_string()],
        last_reindex_vocab: 10,
        consolidated: BTreeMap::new(),
        cycle_count: 1,
        co_recall_counts: BTreeMap::new(),
        last_theta_run_epoch_secs: None,
        last_beta_run_epoch_secs: None,
        last_omega_run_epoch_secs: None,
    };
    first.save_daemon_state(state.clone());

    // A brand-new store over the same backing store (the "restart") sees it.
    let second = EstateManifestDreamingPolicyStore::new(Arc::clone(&s));
    assert_eq!(second.load_daemon_state(), Some(state));
}
